app/src/ingestion/pdf_postprocess.py

```diff
@@ -223,6 +223,59 @@ def _should_merge_text_split_across_pages(text: EnrichedText, next_text: Enriche
     return next_text.text.strip()[0].islower()
 
 
+def _find_heading_index(heading_list: list[str], heading_text: str) -> int:
+    try:
+        return heading_list.index(heading_text)
+    except ValueError:
+        return -1
+
+
+def _group_headings_text(markdown_texts: list[EnrichedText]) -> list[EnrichedText]:
+    grouped_texts_by_headings: list[EnrichedText] = []
+    first_paragraph = markdown_texts[0]
+    combined_text = ""
+    current_heading = first_paragraph.headings
+    page_num = first_paragraph.page_number
+    elem_type = first_paragraph.type
+    headings_processed: list[str] = []
+    for ind, markdown_text in enumerate(markdown_texts):
+        flat_heading = " ".join([h.title for h in current_heading])
+        current_flat_heading = " ".join([h.title for h in markdown_texts[ind].headings])
+        found_prev_heading = _find_heading_index(headings_processed, current_flat_heading)
+        if flat_heading == current_flat_heading:
+            combined_text += f"\n{markdown_texts[ind].text}"
+        else:
+            # only concat the first heading text found if there's a match
+            if found_prev_heading > -1:
+                grouped_texts_by_headings[
+                    found_prev_heading
+                ].text += f"\n{markdown_texts[ind].text}"
+            else:
+                grouped_texts_by_headings.append(
+                    EnrichedText(
+                        text=combined_text.strip(),
+                        type=elem_type,
+                        headings=current_heading,
+                        page_number=page_num,
+                        stylings=markdown_text.stylings,
+                        links=markdown_text.links,
+                    )
+                )
+                combined_text = markdown_text.text
+                page_num = markdown_text.page_number
+                current_heading = markdown_text.headings
+                elem_type = markdown_text.type
+                headings_processed.append(flat_heading)
+        if ind == (len(markdown_texts) - 1):
+            if found_prev_heading > 0:
+                grouped_texts_by_headings[
+                    found_prev_heading
+                ].text += f"\n{markdown_texts[ind].text}"
+            else:
+                grouped_texts_by_headings.append(markdown_text)
```

Comment:
I believe this only runs once and only on the last element. IMHO: it's easier to read if this was moved outside the for-loop and had some code comment to say why it's needed or what it's doing.